designs/template_generation_through_frameworks.md

```diff
@@ -0,0 +1,145 @@
+
+What is the problem?
+--------------------
+
+Customers have different ways to define their AWS Resources. As of writing (Jan. 2109),
+SAM CLI supports the use case of defining an application in CloudFormation/SAM (a super
+set of CloudFormation). These CloudFormation/SAM applications are written in `json` or `yaml`
+and deployed through AWS CloudFormation. Frameworks like CDK offer customers an alternative
+in how they define their applications. SAM CLI should support the ability to invoke functions 
+defined through these other frameworks to enable them to locally debug or manage their 
+applications.
+
+What will be changed?
+---------------------
+
+To start, we will add support for processing metadata from CDK applications:
+SAM CLI will add a processing step on the templates it reads. This will consist of reading
+the template and for each resource reading the metadata and replacing values as specified.
+
+In the future, we can support creating these templates from the different frameworks in a command directly within 
+SAM CLI but is out of scope in the initial implementation of support.
+
+Success criteria for the change
+-------------------------------
+
+* Ability to invoke functions locally that was defined in AWS Cloud Development Kit (CDK).
+* Process a template with CDK Metadata on a resource.
+
+Out-of-Scope
+------------
+
+* A command that will generate the template from the framework.
+* Handling multiple stacks.
+* Support for frameworks other than CDK.
+
+User Experience Walkthrough
+---------------------------
+
+### Customer using CDK
+
+A customer will use CDK to generate the template. This can be done by generating a template and saving it to a file:
+`cdk synth > template.yaml`. Then will then be able to `sam local [invoke|start-api|start-lambda]` any 
+function they have defined [1].   
+
+
+[1] Note: The cdk version must be greater than v0.21.0 as the metadata needed to parse is not appended on older versions. 
+
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+For the features currently in scope, there are no changes to the CLI interface.
+
+### Breaking Change
+
+No breaking changes
+
+Design
+------
+
+All the providers, which are used to get resources out of the template provided to the command, call 
+`SamBaseProvider.get_template(template_dict, parameter_overrides)` to get a normalized template. This function call is 
+responsible for taking a SAM template dictionary and returning a cleaned copy of the template where SAM plugins have 
+been run and parameter values have been substituted. Given the current scope of this call, expanding it to also normalize 
+metadata, seems reasonable. We will expand `SamBaseProvider.get_tempalte()` to call a `CdkTemplateNormalizer` class
+that will be responsible for understanding the metadata and normalizing the template with respect to the metadata. 
+
+```python
+
+class CdkTemplateNormalizer(object):
```

Comment:
Thinking about this more. I think this should be a metadata normalized instead of tying it to CDK. CDK happens to be the first to add this type of metadata but nothing here forces it to be from CDK. 

I will work on removing references to that directly and center the design around the Metadata instead. I will also add a snippet of the Metadata on a Resource (slipped my mind when initially writing)

Reply:
+1 we designed the metadata keys (aws:asset:*) to not be tied directly to cdk with the purpose of allowing other frameworks to utilize the same approach.